### PR TITLE
Refactoring suggestions

### PR DESCRIPTION
**Pull request by kalemas**
Hey @krathjen, what do you think of making it with functional approach?

I m a fan of functional programming and cross platform architectures )) I cant agree with you to delay such kind of changes, we could talk about advantages/disadvantages of those two things next time we will have voice conversation.

This way we still allow client to add own LibraryWidget implementation and we are not forced to rewrite whole project's code. This thing would live separately but all new features would be implemented with regard of this solution.

### CODE REVIEW

**Comment by krathjen** [comment]
Hey @kalemas 

Thanks for looking at this.

This approach is good, however, at the moment it seems to add unnecessary complexity.

I would prefer one of the following to keep things simple.

Kurt


```python
# ----------------------------------
# Example 1
# ----------------------------------

# Maya Shelf button

import studiolibrarymaya
w = studiolibrarymaya.MayaLibraryWidget.instance(*args, **kwargs)
w.show()
```
```python
# ----------------------------------
# Example 2
# ----------------------------------

# studiolibrarymaya/main.py

def main(*args, **kwargs):

    studiolibrarymaya.registerItems()
    
    w = studiolibrarymaya.MayaLibraryWidget.instance(*args, **kwargs)
    w.show()

    return w


#  Maya shelf button

import studiolibrarymaya
studiolibrarymaya.main()
```

**Comment by kalemas** [comment]
So actually public code with this solution will be:
``` python
# ----------------------------------
# Example 1
# ----------------------------------
import studiolibrary
studiolibrary.main()
```
no matter where you run it: Maya, Max, Houdini, Blender. As for now in Max it just run as it is, no items but dialog with folder structure is opened.

If you need custom widget or any custom logic you have to override Platform:
``` python
# ----------------------------------
# Example 2
# ----------------------------------
import studiolibrary.platform
import studiolibrarymaya

class CustomMayaPlatform(studiolibrary.platform.Platform):
    def __init__(self):
        super(CustomMayaPlatform, self).__init__()
        studiolibrarymaya.registerItems()

    def mainWidgetClass(self, _):
        return studiolibrarymaya.MayaLibraryWidget

studiolibrary.main(platform=CustomMayaPlatform())
```
yes it is more complex, but:
- at moment we haven't any organized interface to extend it, so third party developer have to do fork or invent own wheel. It is always better to build on top rather than to fork.
- it is functional style (no globals, no side effects, debugging is easier, use functions as tools, etc.)
- maybe i will remember anything else ))

**Comment by kalemas** [comment]
- when you mind in terms of "building on top" you will came to better architecture because you will not do what you want and get data in in managed way. You will organize you requests and procedures through mediator object, found something unnecessary and more.

I count each modern application have to be based on something like that, but i realize that it is only my opinion...

**Comment by krathjen** [comment]
Hey @kalemas 

The revised Platform class looks nicer, however, I still think it's not needed at this point.

I would like to try and reduce the code more and avoid adding too many levels of abstraction.

What is the reason for extending the LibraryWidget class for Max?

I would still prefer an interface similar to the following at the moment.

```python
import studiolibrary


class CustomLibraryWidget(studiolibrary.LibraryWidget)
    pass


studiolibrary.main(cls=CustomLibraryWidget)
```

**Comment by kalemas** [comment]
Your example is much better, platform dependent logic isolated from client, although it still doesn't localize platform dependent code so it would not be used inside or by third party in unified manner, like this:

``` python
class BaseAnimItem(BaseItem):
    def make(self):
        super(BaseAnimItem, self).make(thumbnailPath=self.parent().platform().grabMovie())

class BaseScriptItem(BaseItem):
    def apply(self):
        self.parent().platform().execute(self.data())
```
Sorry i still doesn't get deep into your items organization so all names are fictitious )

Hm, maybe at the back of my mind i want to make cross DCC operation library.

> What is the reason for extending the LibraryWidget class for Max?

At this point i didn't made any changes for Max, however you have to inherit from specific class to enable interface docking.

**Comment by krathjen** [comment]
yea, I would like to change a few things in the LibraryItem. The LibraryItem.save method needs to be simplified as well as the items in the studiolibrarymaya package.

If you have time maybe have a look at the example in "studiolibrary/docs/examples/examples1.py". You should be able to just run the file.

I'll also have a think about the item structure a bit more. I would probably start with creating the MaxItems independent of the studiolibrarymaya package and see how they involve.

Btw, I'm away next week, so I might be a bit quiet. ;)
    
```python
import maxlib
import studiolibrary


class MaxAnimItem(studiolibrary.Library):

    def load(self):
       maxlib.loadAnim(self.path(), selection=True)

    def save(self):
       maxlib.saveAnim(self.path(), selection=True)
        

studiolibrary.registerItem(MaxAnimItem, ".maxanim")
```

**Comment by kalemas** [comment]
So i getting deeper )) 

- i found that you catching `ImportError` on `import maya.cmds` in `mutils`, i count that it would be fine for development but do nothing useful for production. What your use cases are for that?
- item registration isn't functional-style. I d prefer to get rid of global variables almost completely.

**Comment by kalemas** [comment]
> `maxlib`

You may want to implement platform support based on python modules, i realized that even with some kind of inheritance it may do it haven't such power as classes. Most of DCCs have a lot of similar logic that differs in few values, with classes you could organize complex interdependent architecture to get rid of duplicated logic.

I personally do not fear complex structures and flexibility, i fear more non-trivial solutions and endless refactoring. Overriding classes to change behavior is quite clear for me.

With functional style you remove unnecessary dependencies and usual have not so much solutions (that are all good). Also you have to displace logic to entities that use it.

**Comment by kalemas** [comment]
> If you have time maybe have a look at the example in "studiolibrary/docs/examples/examples1.py". You should be able to just run the file.

Nice example )) I only suggest to change item registration like in b409e02 , by the way we may run studiolibrary with different setup in the same environment (for example, Pose Library and Movie Library from your example). I would not realize some things (for if you use registered items somewhere else but LibraryWidget) so those refactoring suggestions help me to realize it deeper )

**Comment by kalemas** [comment]
So tell me, why you don't like it? ))))

Using function is flexible and simple for filtering in `FileSystemModel` and we are free to build on top of it anything with not been affected of already registered items.

**Comment by krathjen** [comment]
Hey @kalemas,

I was away last week and I'm pretty busy for the next week. 

I'll try and have a look at this soon. 

Thanks,

Kurt

**Comment by krathjen** [comment]
Hey @kalemas,

Sorry, August has been a very busy month.

You have been busy! :) As there are many commits it might be best to break these up into separate pull requests. It should make it easier to review.

I'll try and comment on each commit for now.

Thanks,

Kurt

File: docs/examples/example1.py [hunk 1]
```diff
@@ -39,6 +39,16 @@ def previewWidget(self, libraryWidget):
         """
         return ImagePreviewWidget(self)
 
+    @staticmethod
+    def isPathSuitable(self, path):
```
Comment (krathjen) [inline docs/examples/example1.py]: This looks good. 

It might be better to name the method to "isPathValid".  This will make the name more consistent with the QT naming style.

Also, can we make this change a separate pull request?

Thanks

File: docs/examples/example1.py [hunk 2]
```diff
@@ -87,17 +97,20 @@ def resizeEvent(self, event):
 def main():
     """The main entry point for this example."""
 
-    # Register the item class to the extension and ignore all jpgs that
-    # contain the string "thumbnail."
-    studiolibrary.registerItem(ImageItem, ".jpg", ignore="thumbnail.")
-    studiolibrary.registerItem(ImageItem, ".png")
-    studiolibrary.registerItem(ImageItem, ".gif")
+    class Widget(studiolibrary.LibraryWidget):
```
Comment (krathjen) [inline docs/examples/example1.py]: I like the idea of having the option to reimplement the "itemClasses", however, I'm not sure it should be necessary to create a LibraryWidget to register items.

Also, we should probably change the name to something like "registeredItems()", because the setter is "registerItem(itemClass)". What do you think?

We should also make this improvement a separate pull request.

Example:

```python
# /studiolibrary/librarywidget.py
class LibraryWidget(QWidget):
    ...
    def registeredItems(self):
        return studiolibrary.registeredItems()
```

File: main.py [hunk 3]
```diff
@@ -35,12 +38,23 @@ def main(
     :type superusers: str
     :type lockFolder: str
     :type unlockFolder: str
```
Comment (krathjen) [inline main.py]: I still need to think about this change a bit more, however, I would like to have the main function just for convenience and encourage developers/studio's to call their custom library widget directly.  We should probably make this change a separate pull request too.

Example:

```python
import customlibrarywidget
w = customlibrarywidget.CustomLibraryWidget.instance()
```

It would also be nice to avoid a "studiolibrarymaya" dependency in the studiolibrary code.

I'll need to think about this more, however, I was thinking something like the following.

```python
def main(cls=None, **kwargs):
    """
    Convenience method for creating a libraryWidget instance.
    """
    cls = cls or studiolibrary.LIBRARY_WIDGET_CLASS or studiolibrary.LibraryWidget
    return cls.instance(**kwargs)
```

To support the current main function we would still need to do something like the following for now.

```python
#studiolibrary/config.py

if studiolibrary.isMaya():
    import studiolibrarymaya.mayalibrarywidget
    studiolibrary.LIBRARY_WIDGET_CLASS = studiolibrarymaya.mayalibrarywidget.MayaLibraryWidget
```

Happy to discuss this further. :)

**Comment by kalemas** [comment]
@krathjen i see one important point here before closing this request. It is platform support.

1. Well, suppose i would like to extend StudioLiabrary to support Max, i made own widget that mixin MaxDockable, but i need any sort of initialization outside of that widget to register items.

I count that information about classes widget can handle belongs to widget not to the application.

2. Well StudioLibrary support Maya out of the box, is it supposed that any other application will be supported in the same way? If so, you will expose MayaLibraryWidget and similar classes to client?

I count any new environment would be supported by custom widget, but then it have to be encapsulated inside StudioLibrary, so behavior for high level classes should be same in diferent environments.

**Comment by krathjen** [comment]
Hey @kalemas,

Great to hear from you!

I see your point about Studio Library supporting Maya out of the box.  I guess we would need to change the config.py or main.py files to one of the following to also support Max out of the box. Similar to what you have suggested.



At the moment you can only register new item types on a global level. I have created a ticket to look at support for this on the LibraryWidget level.

Have you created new item types for max?

```python
# Example 1
# studiolibrary/config.py
if studiolibrary.isMax():
    import studiolibrarymax
    libraryWidget = studiolibrarymax.setup()
elif studiolibrary.isMaya():
    import studiolibrarymaya
    libraryWidget = studiolibrarymaya.setup()
```
```python
# Example 2
# studiolibrary/main.py
def main(cls=None, *args, **kwargs):
    """
    Convenience method for creating and loading a library widget instance.

    :type cls: studiolibrary.LibraryWidget.__class__ or None
    :rtype: studiolibrary.LibraryWidget
    """
    cls = cls or studiolibrary.LibraryWidget

    if studiolibrary.isMax():
        import studiolibrarymax
        libraryWidget = studiolibrarymax.main(*args, **kwargs)

    elif studiolibrary.isMaya():
        import studiolibrarymaya
        libraryWidget = studiolibrarymaya.main(*args, **kwargs)

    else:
        libraryWidget = cls.instance(*args, **kwargs)

    return libraryWidget
```

**Comment by krathjen** [comment]
@kalemas do you mind if we close this pull request?

**Comment by kalemas** [comment]
You suggest to object platform abstraction layer and provide separate interfaces for each platform.  I can't agree with you about we would close it.

Get rid of globals make sense, with python you could use not global/module level variables at all. I will look at that ticket.

We do not created any items for Max as we completely move on Maya.

**Comment by krathjen** [comment]
Thanks, @kalemas!

Also, thanks for fixing the lock folder bug! 👍
